File: tradebot/data/ohlcv.py
from __future__ import annotations

import asyncio
from dataclasses import dataclass
from datetime import UTC, datetime

from tradebot.data.prices import PriceTick
from tradebot.storage.repo import JsonStorage, OHLCVCandle

TIMEFRAME_SECONDS: dict[str, int] = {
    "5s": 5,
    "1m": 60,
    "15m": 900,
    "1h": 3600,
}


def bucket_for(ts: datetime, timeframe: str) -> datetime:
    secs = TIMEFRAME_SECONDS[timeframe]
    epoch = int(ts.replace(tzinfo=UTC).timestamp())
    floored = epoch - (epoch % secs)
    return datetime.fromtimestamp(floored, tz=UTC)


@dataclass
class _BucketState:
    bucket_start: datetime
    open: float
    high: float
    low: float
    close: float


class OHLCVAggregator:
    def __init__(
        self, storage: JsonStorage, source: asyncio.Queue[PriceTick], timeframe: str
    ) -> None:
        self._storage = storage
        self._source = source
        self._timeframe = timeframe
        self._state: dict[str, _BucketState] = {}
        self._stop_after_drain = False

    def stop_after_drain(self) -> None:
        self._stop_after_drain = True

    async def run(self) -> None:
        while True:
            try:
                tick = await asyncio.wait_for(self._source.get(), timeout=0.05)
            except TimeoutError:
                if self._stop_after_drain:
                    return
                continue
            await self._handle(tick)

    async def _handle(self, tick: PriceTick) -> None:
        bucket = bucket_for(tick.sampled_at, self._timeframe)
        cur = self._state.get(tick.pair)
        if cur is None:
            state = _BucketState(bucket, tick.price, tick.price, tick.price, tick.price)
            self._state[tick.pair] = state
            return
        if bucket > cur.bucket_start:
            self._write(tick.pair, cur)
            state = _BucketState(bucket, tick.price, tick.price, tick.price, tick.price)
            self._state[tick.pair] = state
            return
        cur.high = max(cur.high, tick.price)
        cur.low = min(cur.low, tick.price)
        cur.close = tick.price

    def _write(self, pair: str, state: _BucketState) -> None:
        self._storage.upsert_ohlcv(
            OHLCVCandle(
                pair=pair,
                timeframe=self._timeframe,
                bucket_start=state.bucket_start,
                open=state.open,
                high=state.high,
                low=state.low,
                close=state.close,
                volume_quote=0.0,  # populated in a later phase when fills exist
            ),
        )

    def _flush_all(self) -> None:
        for pair, state in self._state.items():
            self._write(pair, state)
        self._state.clear()

File: crates/data/src/ohlcv.rs
//! OHLCV candle aggregator. Port of `tradebot/data/ohlcv.py`.

use std::collections::HashMap;
use std::time::Duration;

use chrono::{DateTime, TimeZone, Utc};
use rust_decimal::Decimal;
use tokio::sync::mpsc;
use tradebot_storage::{JsonStorage, OhlcvCandle, StorageError};

use crate::prices::PriceTick;

/// Bucket width in seconds for each supported timeframe. Matches
/// `TIMEFRAME_SECONDS` in `tradebot/data/ohlcv.py`.
pub fn timeframe_seconds(timeframe: &str) -> Option<i64> {
    match timeframe {
        "5s" => Some(5),
        "1m" => Some(60),
        "15m" => Some(900),
        "1h" => Some(3600),
        _ => None,
    }
}

/// Floor `ts` to the start of its bucket for `timeframe`.
///
/// # Panics
/// Panics if `timeframe` is not one of the supported values, matching the
/// Python `KeyError` from indexing `TIMEFRAME_SECONDS`.
pub fn bucket_for(ts: DateTime<Utc>, timeframe: &str) -> DateTime<Utc> {
    let secs =
        timeframe_seconds(timeframe).unwrap_or_else(|| panic!("unknown timeframe: {timeframe}"));
    let epoch = ts.timestamp();
    let floored = epoch - epoch.rem_euclid(secs);
    Utc.timestamp_opt(floored, 0)
        .single()
        .expect("floored timestamp is always representable")
}

#[derive(Debug, Clone)]
struct BucketState {
    bucket_start: DateTime<Utc>,
    open: f64,
    high: f64,
    low: f64,
    close: f64,
}

fn f64_to_money(value: f64) -> Decimal {
    Decimal::from_f64_retain(value).unwrap_or_default()
}

/// Consumes `PriceTick`s from a channel and writes one OHLCV candle per
/// timeframe bucket per pair to `storage`.
pub struct OhlcvAggregator<'a> {
    storage: &'a JsonStorage,
    source: mpsc::Receiver<PriceTick>,
    timeframe: String,
    state: HashMap<String, BucketState>,
    stop_after_drain: bool,
}

impl<'a> OhlcvAggregator<'a> {
    pub fn new(
        storage: &'a JsonStorage,
        source: mpsc::Receiver<PriceTick>,
        timeframe: impl Into<String>,
    ) -> Self {
        Self {
            storage,
            source,
            timeframe: timeframe.into(),
            state: HashMap::new(),
            stop_after_drain: false,
        }
    }

    pub fn stop_after_drain(&mut self) {
        self.stop_after_drain = true;
    }

    /// Drain ticks from `source`, writing a candle each time a pair's bucket
    /// rolls over. Returns once `stop_after_drain` has been set and no tick
    /// arrives within the poll window, or once the channel is permanently
    /// closed (no `PriceFeed` side left to send more ticks).
    pub async fn run(&mut self) -> Result<(), StorageError> {
        loop {
            match tokio::time::timeout(Duration::from_millis(50), self.source.recv()).await {
                Ok(Some(tick)) => self.handle(tick)?,
                Ok(None) => return Ok(()),
                Err(_elapsed) => {
                    if self.stop_after_drain {
                        return Ok(());
                    }
                }
            }
        }
    }

    fn handle(&mut self, tick: PriceTick) -> Result<(), StorageError> {
        let bucket = bucket_for(tick.sampled_at, &self.timeframe);
        match self.state.get_mut(&tick.pair) {
            None => {
                self.state.insert(
                    tick.pair,
                    BucketState {
                        bucket_start: bucket,
                        open: tick.price,
                        high: tick.price,
                        low: tick.price,
                        close: tick.price,
                    },
                );
                Ok(())
            }
            Some(cur) if bucket > cur.bucket_start => {
                let finished = cur.clone();
                self.write(&tick.pair, &finished)?;
                self.state.insert(
                    tick.pair,
                    BucketState {
                        bucket_start: bucket,
                        open: tick.price,
                        high: tick.price,
                        low: tick.price,
                        close: tick.price,
                    },
                );
                Ok(())
            }
            Some(cur) => {
                cur.high = cur.high.max(tick.price);
                cur.low = cur.low.min(tick.price);
                cur.close = tick.price;
                Ok(())
            }
        }
    }

    fn write(&self, pair: &str, state: &BucketState) -> Result<(), StorageError> {
        let candle = OhlcvCandle {
            pair: pair.to_string(),
            timeframe: self.timeframe.clone(),
            bucket_start: state.bucket_start,
            open: f64_to_money(state.open),
            high: f64_to_money(state.high),
            low: f64_to_money(state.low),
            close: f64_to_money(state.close),
            volume_quote: Decimal::ZERO,
        };
        self.storage.upsert_ohlcv(&candle)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bucket_for_floors_to_minute() {
        let ts = Utc.with_ymd_and_hms(2026, 5, 3, 12, 0, 30).unwrap();
        let bucket = bucket_for(ts, "1m");
        assert_eq!(bucket, Utc.with_ymd_and_hms(2026, 5, 3, 12, 0, 0).unwrap());
    }

    #[test]
    fn bucket_for_floors_to_hour() {
        let ts = Utc.with_ymd_and_hms(2026, 5, 3, 12, 45, 10).unwrap();
        let bucket = bucket_for(ts, "1h");
        assert_eq!(bucket, Utc.with_ymd_and_hms(2026, 5, 3, 12, 0, 0).unwrap());
    }
}
